fsgrid.hpp

```diff
@@ -512,17 +512,125 @@ template <typename T, int stencil> class FsGrid : public FsGridTools{
          this->data = other.getData(); // Copy assignment
       }
 
-      /*! Finalize instead of destructor, as the MPI calls fail after the main program called MPI_Finalize().
-       *  Cleans up the cartesian communicator
+      /*! 
+       *  MPI calls fail after the main program called MPI_Finalize(),
+       *  so this can be used instead of the destructor
+       *  Cleans up the cartesian communicator and datatypes
        */
-      void finalize() {
-         for(int i=0;i<27;i++){
-            if(neighbourReceiveType[i] != MPI_DATATYPE_NULL)
-               MPI_Type_free(&(neighbourReceiveType[i]));
-            if(neighbourSendType[i] != MPI_DATATYPE_NULL)
-               MPI_Type_free(&(neighbourSendType[i]));
+      void finalize() noexcept {
+         if (comm3d != MPI_COMM_NULL) {
+            MPI_Comm_free(&comm3d);
+            comm3d = MPI_COMM_NULL;
+         }
+         for (auto& type : neighbourReceiveType) {
+            if (type != MPI_DATATYPE_NULL) {
+               MPI_Type_free(&type);
+               type = MPI_DATATYPE_NULL;
+            }
+         }
+         for (auto& type : neighbourSendType) {
+            if (type != MPI_DATATYPE_NULL) {
+               MPI_Type_free(&type);
+               type = MPI_DATATYPE_NULL;
+            }
+         }
+      }
+
+      /*!
+       *  If finalize() isn't called manually, object should be destroyed before MPI finalization
+       */
+      ~FsGrid() {
+         finalize();
+      }
+
+      friend void swap (FsGrid& first, FsGrid& second) noexcept {
```

Comment:
Why does this need to be declared ```friend```, if we have no private or protected members in the fsgrid struct?
(My personal rule of thumb is: if you have to write "friend" on a function, you are probably overusing C++ object member protection features)

Reply:
Only `DX`, `DY`, `DZ`, and `physicalGlobalStart` are public, the rest are private. This particular implementation of swap is apparently [idiomatic](https://stackoverflow.com/questions/5695548/public-friend-swap-member-function).

Reply:
Alright!